### PR TITLE
cmd/snap-preseed: use snapd from the deb if newer than from seeds

### PR DESCRIPTION
**Pull request by stolowski**
Compare snapd version from the deb and core*/snapd snap when preseeding, and use the newer one.

### CODE REVIEW

**Comment by anonymouse64** [comment]
seems gofmt is unhappy:

```
2020-08-04T08:58:40.7068529Z Checking formatting
2020-08-04T08:58:40.7068634Z Formatting wrong in following files:
2020-08-04T08:58:40.7068975Z /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/snap-preseed/main_test.go
2020-08-04T08:58:40.7069395Z diff -u /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/snap-preseed/main_test.go.orig /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/snap-preseed/main_test.go
2020-08-04T08:58:40.7069783Z --- /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/snap-preseed/main_test.go.orig	2020-08-04 08:58:37.734163596 +0000
2020-08-04T08:58:40.7070156Z +++ /tmp/static-unit-tests/src/github.com/snapcore/snapd/cmd/snap-preseed/main_test.go	2020-08-04 08:58:37.734163596 +0000
2020-08-04T08:58:40.7070411Z @@ -424,47 +424,47 @@
2020-08-04T08:58:40.7070535Z  	restoreMountPath := main.MockSnapdMountPath(targetSnapdRoot)
2020-08-04T08:58:40.7070652Z  	defer restoreMountPath()
2020-08-04T08:58:40.7070759Z  
2020-08-04T08:58:40.7071202Z -	var versions = []struct{
2020-08-04T08:58:40.7071467Z -		fromSnap string
2020-08-04T08:58:40.7071695Z -		fromDeb string
2020-08-04T08:58:40.7071907Z -		expectedPath string
2020-08-04T08:58:40.7072018Z +	var versions = []struct {
2020-08-04T08:58:40.7072126Z +		fromSnap        string
2020-08-04T08:58:40.7072232Z +		fromDeb         string
2020-08-04T08:58:40.7072335Z +		expectedPath    string
2020-08-04T08:58:40.7072437Z  		expectedVersion string
2020-08-04T08:58:40.7072661Z -		expectedErr string
2020-08-04T08:58:40.7072751Z +		expectedErr     string
2020-08-04T08:58:40.7072855Z  	}{
2020-08-04T08:58:40.7072964Z  		{
2020-08-04T08:58:40.7073191Z -			fromDeb: "2.44.0",
2020-08-04T08:58:40.7073298Z +			fromDeb:  "2.44.0",
2020-08-04T08:58:40.7073405Z  			fromSnap: "2.45.3+git123",
2020-08-04T08:58:40.7073510Z  			// snap version wins
2020-08-04T08:58:40.7073739Z  			expectedVersion: "2.45.3+git123",
2020-08-04T08:58:40.7074054Z -			expectedPath: filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7074895Z +			expectedPath:    filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7075053Z  		},
2020-08-04T08:58:40.7075155Z  		{
2020-08-04T08:58:40.7075433Z -			fromDeb: "2.44.0",
2020-08-04T08:58:40.7075541Z +			fromDeb:  "2.44.0",
2020-08-04T08:58:40.7075630Z  			fromSnap: "2.44.0",
2020-08-04T08:58:40.7075749Z  			// snap version wins
2020-08-04T08:58:40.7075900Z  			expectedVersion: "2.44.0",
2020-08-04T08:58:40.7076279Z -			expectedPath: filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7076610Z +			expectedPath:    filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7076731Z  		},
2020-08-04T08:58:40.7076833Z  		{
2020-08-04T08:58:40.7077072Z -			fromDeb: "2.45.1+20.04",
2020-08-04T08:58:40.7077172Z +			fromDeb:  "2.45.1+20.04",
2020-08-04T08:58:40.7077281Z  			fromSnap: "2.45.1",
2020-08-04T08:58:40.7077382Z  			// deb version wins
2020-08-04T08:58:40.7077491Z  			expectedVersion: "2.45.1+20.04",
2020-08-04T08:58:40.7080987Z -			expectedPath: filepath.Join(tmpDir, "usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7081133Z +			expectedPath:    filepath.Join(tmpDir, "usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7081244Z  		},
2020-08-04T08:58:40.7081327Z  		{
2020-08-04T08:58:40.7081559Z -			fromDeb: "2.45.2",
2020-08-04T08:58:40.7081663Z +			fromDeb:  "2.45.2",
2020-08-04T08:58:40.7081765Z  			fromSnap: "2.45.1",
2020-08-04T08:58:40.7082294Z  			// deb version wins
2020-08-04T08:58:40.7082399Z  			expectedVersion: "2.45.2",
2020-08-04T08:58:40.7082674Z -			expectedPath: filepath.Join(tmpDir, "usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7083028Z +			expectedPath:    filepath.Join(tmpDir, "usr/lib/snapd/snapd"),
2020-08-04T08:58:40.7083140Z  		},
2020-08-04T08:58:40.7083224Z  		{
2020-08-04T08:58:40.7083464Z -			fromSnap: "2.45.1",
2020-08-04T08:58:40.7083578Z +			fromSnap:    "2.45.1",
2020-08-04T08:58:40.7083709Z  			expectedErr: fmt.Sprintf("cannot open snapd info file %q.*", filepath.Join(tmpDir, "usr/lib/snapd/info")),
2020-08-04T08:58:40.7083832Z  		},
2020-08-04T08:58:40.7083929Z  		{
2020-08-04T08:58:40.7084152Z -			fromDeb: "2.45.1",
2020-08-04T08:58:40.7084258Z +			fromDeb:     "2.45.1",
2020-08-04T08:58:40.7084599Z  			expectedErr: fmt.Sprintf("cannot open snapd info file %q.*", filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/info")),
2020-08-04T08:58:40.7084725Z  		},
2020-08-04T08:58:40.7084823Z  	}
2020-08-04T08:58:40.7084890Z 
2020-08-04T08:58:40.7084938Z 
2020-08-04T08:58:40.7085041Z Crushing failure and despair.
```

File: cmd/snap-preseed/main_test.go [hunk 1]
```diff
@@ -398,19 +405,91 @@ func (s *startPreseedSuite) TestRunPreseedUnsupportedVersion(c *C) {
...
+		{
+			fromDeb:  "2.44.0",
+			fromSnap: "2.44.0",
+			// snap version wins
+			expectedVersion: "2.44.0",
+			expectedPath:    filepath.Join(tmpDir, "target-core-mounted-here/usr/lib/snapd/snapd"),
+		},
+		{
+			fromDeb:  "2.45.1+20.04",
+			fromSnap: "2.45.1",
```
Comment (anonymouse64) [inline cmd/snap-preseed/main_test.go:450]: thanks for this

File: cmd/snap-preseed/main_test.go [hunk 2]
```diff
@@ -373,6 +379,7 @@ func (s *startPreseedSuite) TestClassicRequired(c *C) {
 func (s *startPreseedSuite) TestRunPreseedUnsupportedVersion(c *C) {
```
Comment (pedronis) [inline cmd/snap-preseed/main_test.go:379]: shouldn't we have a TestRunPreseedDebIsNewer as well?

File: cmd/snap-preseed/preseed_linux.go [hunk 3]
```diff
@@ -150,21 +150,52 @@ var systemSnapFromSeed = func(rootDir string) (string, error) {
...
 
-func checkTargetSnapdVersion(infoPath string) error {
-	ver, err := snapdtool.SnapdVersionFromInfoFile(infoPath)
+// chooseTargetSnapdVersion checks if the version of snapd under chroot env
+// is good enough for preseeding. It checks both the snapd from the deb
+// and from the seeded snap mounted under snapdMountPath and returns the
+// full path of snapd to execute as part of preseeding (whichever version is
+// newer).
+// The function must be called after syscall.Chroot(..).
+func chooseTargetSnapdVersion() (snapdPath string, whichVer string, err error) {
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go:159]: this should return a struct I think

File: cmd/snap-preseed/preseed_linux.go [hunk 4]
```diff
@@ -215,14 +246,6 @@ func prepareChroot(preseedChroot string) (func(), error) {
 		}
 	}
 
-	// read version from the mounted core snap
-	infoPath := filepath.Join(snapdMountPath, dirs.CoreLibExecDir, "info")
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go:219]: with this change we fail later now, that seems a bit annoying.  can't we pass the directory as seen from outside to choose... ? then choose return a struct and we pass the struct to runPreseed ?

File: cmd/snap-preseed/preseed_linux.go [hunk 5]
```diff
@@ -150,21 +150,52 @@ var systemSnapFromSeed = func(rootDir string) (string, error) {
...
+// chooseTargetSnapdVersion checks if the version of snapd under chroot env
+// is good enough for preseeding. It checks both the snapd from the deb
+// and from the seeded snap mounted under snapdMountPath and returns the
+// full path of snapd to execute as part of preseeding (whichever version is
+// newer).
+// The function must be called after syscall.Chroot(..).
+func chooseTargetSnapdVersion() (snapdPath string, whichVer string, err error) {
+	// read snapd version from the mounted core/snapd snap
+	infoPath := filepath.Join(snapdMountPath, dirs.CoreLibExecDir, "info")
+	ver1, err := snapdtool.SnapdVersionFromInfoFile(infoPath)
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go:162]: verFromSnap perhaps?

File: cmd/snap-preseed/preseed_linux.go [hunk 6]
```diff
@@ -150,21 +150,52 @@ var systemSnapFromSeed = func(rootDir string) (string, error) {
...
 	if err != nil {
-		return err
+		return "", "", err
 	}
 
-	res, err := strutil.VersionCompare(ver, snapdPreseedSupportVer)
+	// read snapd version from the main fs under chroot (snapd from the deb);
+	// assumes running under chroot already.
+	infoPath = filepath.Join(dirs.GlobalRootDir, dirs.CoreLibExecDir, "info")
+	ver2, err := snapdtool.SnapdVersionFromInfoFile(infoPath)
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go:170]: debVer ?

File: cmd/snap-preseed/preseed_linux.go (hunk 4 above)
Comment (pedronis) [reply cmd/snap-preseed/preseed_linux.go:219]: yet, another approach  is to leave Check... as it is,  and simply do the second comparision only between snap and deb in runPreseed. If the deb is older than the snap we don't care either way.

File: cmd/snap-preseed/main_test.go (hunk 2 above)
Comment (stolowski) [reply cmd/snap-preseed/main_test.go:441]: Added.

File: cmd/snap-preseed/preseed_linux.go (hunk 4 above)
Comment (stolowski) [reply cmd/snap-preseed/preseed_linux.go:219]: I liked the first idea more, did that (except for passing the directory.. bit), I hope it makes sense?

File: cmd/snap-preseed/preseed_linux.go [hunk 7]
```diff
@@ -150,30 +150,68 @@ var systemSnapFromSeed = func(rootDir string) (string, error) {
...
+type targetSnapdInfo struct {
+	path    string
+	version string
+}
+
+// chooseTargetSnapdVersion checks if the version of snapd under chroot env
+// is good enough for preseeding. It checks both the snapd from the deb
+// and from the seeded snap mounted under snapdMountPath and returns the
+// full path of snapd to execute as part of preseeding (whichever version is
+// newer).
```
Comment (pedronis) [inline cmd/snap-preseed/preseed_linux.go:162]: this comment is slightly imprecise now because we return more than the path

**Pull request merged**